File: 蟒/problem2-2-2.py
# @file		    problem2-2-2.py
# @author      	ZhaoQicheng(zhaoqicheng2022@163.com/love-Tom)
# @brief        
# @detail       
# @version 	    0.1.0
# @date 		2022-4-18
# @copyright	2022-4-18 by zhaoqicheng.All rights reserved. 
  
import numpy as np
import cv2 as cv 

#定义高斯滤波函数
def GaussianFilter(image, K_size=3, sigma=2): #设置卷积核，方差，不设置会自动生成，默认值分别为3*3，2
    #读出图片属性(长宽高)
    H, W, C = image.shape    
    
    #计算需要填充的0层，并填充
    pad = K_size // 2
    image_changed = np.zeros((H + pad * 2, W + pad * 2, C), dtype=np.float64)  #填充后的image_changed
    image_changed[pad: pad + H, pad: pad + W] = image.copy().astype(np.float64)
    
    #卷积核设置
    K = np.zeros((K_size, K_size), dtype=np.float64)   #初始化
    
    #遍历进行操作
    for x in range(-pad, -pad + K_size):
        for y in range(-pad, -pad + K_size):
            K[y + pad, x + pad] = np.exp( -(x ** 2 + y ** 2) / (2 * (sigma ** 2)))  #二维高斯分布中的指数部分
    K /= (2 * np.pi * sigma * sigma)    #二维高斯分布的系数
    K /= K.sum()
    tmp = image_changed.copy()  #拷贝一份用来操作
    
    #遍历进行卷积运算
    for y in range(H):
        for x in range(W):
            for c in range(C):
                image_changed[pad + y, pad + x, c] = np.sum(K * tmp[y: y + K_size, x: x + K_size, c])#先乘后加
    image_changed = np.clip(image_changed, 0, 255)
    image_changed = image_changed[pad: pad + H, pad: pad + W].astype(np.uint8)   #数据类型转换
    return image_changed

#读取图像
image = cv.imread("D:/photo/lena.jpg")
#对图像进行处理
image_changed = GaussianFilter(image, 3, 2) 
cv.imshow("image",image)
cv.imshow("image_changed",image_changed)      

cv.waitKey(0)
cv.destroyAllWindows()

File: C++/problem2-2-2.cpp
/* @file                problem2-2-2.cpp
 * @author         zhaoqicheng(zhaoqicheng2022@163.com)
 * @brief             z
 * @details         z
 * @version        0.1.0
 * @date             2022-4-11
 * @copyright   2022-4-11 by zhaoqicheng.All rights reserved.
 */

#include<opencv2/opencv.hpp>
#include<iostream>
#include<opencv2/core.hpp>
#include<cmath>
#include<opencv2/highgui.hpp>
using namespace std;
using namespace cv;

//函数声明
void listPixel_Operate(int, void*);
void Gaussian(const Mat &input, Mat &output, int a,int b,double sigma);

int sig;
int a,b;
Mat image;

int main() {
    cout << "请输入sigma的值：" << endl;
    cin >> sig;
    sig = 2;
    cout<<"请分别输入卷积核行列"<<endl;
    cin>>a;
    cin>>b;
    image = imread("/home/jichuan-ftp/image/unchanged/lena.jpg");
    namedWindow("【效果图窗口】", 1);
    //调用回调函数
    listPixel_Operate(sig, 0);
    //system("pause");
    waitKey(0);
    return 0;
}
、
//函数声明
void Gaussian(const Mat &input, Mat &output, int a,int b,double sigma) {
    int row = input.rows, col = input.cols;
   // int window = (int)((6 * sigma - 1) / 2) * 2 + 1;//滤波窗口
    int window = a*b;
    double *temp = new double[window];

    //扩充边界
    Mat INPUT;
    copyMakeBorder(input, INPUT, window / 2, window / 2, window / 2, window / 2, BORDER_REFLECT_101);

    double sum = 0;
    for (int w = 0; w < window; w++) {
        int mid = w - window / 2;
        temp[w] = exp(-(mid*mid) / (2 * sigma*sigma));
        sum += temp[w];
    }

    //归一化滤波核，和为1
    for (int w = 0; w < window; w++)
        temp[w] = temp[w] / sum;

    //扩充边界之后的长宽
    int rows = row + window - 1;
    int cols = col + window - 1;

    //先对每行进行一维高斯滤波
    for (int y = window / 2; y < row + window / 2; y++)//行
    {
        for (int x = window / 2; x < col + window / 2; x++) //列
        {
            int num = 0;
            double pix[3] = { 0 };
            for (int k = x - window / 2; k < x + window / 2; k++)
            {
                for (int c = 0; c < INPUT.channels(); c++)
                {
                    pix[c] += (INPUT.at<Vec3b>(y, k)[c])*temp[num];   //列坐标<矩阵列数
                }
                num++;
            }
            for (int c = 0; c < INPUT.channels(); c++)
            {
                INPUT.at<Vec3b>(y, x)[c] = pix[c];
            }
        }
    }

    //再对每列进行一维高斯滤波
    for (int x = window / 2; x < col + window / 2; x++) {
        for (int y = window / 2; y < row + window / 2; y++) {
            int num = 0;
            double pix[3] = { 0 };
            for (int k = y - window / 2; k < y + window / 2; k++) {
                for (int c = 0; c < INPUT.channels(); c++)
                    pix[c] += (INPUT.at<Vec3b>(k, x)[c])*temp[num];
                num++;
            }
            for (int c = 0; c < INPUT.channels(); c++)
                INPUT.at<Vec3b>(y, x)[c] = pix[c];
        }
    }
    for (int y = 0; y < row; y++) {
        for (int x = 0; x < col; x++)  {
            output.at<Vec3b>(y, x) = INPUT.at<Vec3b>(y + window / 2, x + window / 2);
        }
    }
}


void listPixel_Operate(int, void*) {
    Mat dst = Mat::zeros(image.rows, image.cols, image.type());
    imshow("原图窗口", image);
    Gaussian(image, dst,a,b,sig);
    imshow("【效果图窗口】", dst);
}

